### DISCUSSION TITLE
Spritefile archives in parkobj

### DISCUSSION THREAD

**Discussion by spacek531** [OpenRCT2/OpenRCT2#16746] [Feature requests]
I believe there was some work to decrease load times of .parkobj with uncompressed graphics, specifically freezing when mousing over the object in the object selection window, but the problem still persists. At the time, it was discussed that the major issue was RLE-compressing all the images. It stands to reason that this work could be skipped if the images came pre-RLE-compressed, perhaps in a format OpenRCT2 already uses for lots of graphics like an `rct_gx`.

**Comment by X123M3-256** [comment]
I support this. I think it should be optional (after all, being able to use PNG files is one of the major advantages of the .parkobj format), but for objects with a lot of sprites this would speed up loading.

**Comment by spacek531** [comment]
I have a working prototype now, this train ride is comprised of a new object which pulls its sprites from an included `rct_gx` which I have taken to calling "LGX" for Local gx or "LIA" for Local Image Archive EDIT: or spritefile archive
![image](https://user-images.githubusercontent.com/3454156/156719558-dcb4286a-c887-4302-8214-dd934cccae91.png)

[Feature Branch](https://github.com/spacek531/OpenRCT2/tree/objects-rct-gx)

[Control Object (nrl2 converted to existing parkobj)](https://github.com/OpenRCT2/OpenRCT2/files/8183981/spacek.ride.nrl2-control.zip)

[Test Object (nrl2 converted to use LGX)](https://github.com/OpenRCT2/OpenRCT2/files/8183983/spacek.ride.nrl2-test.zip)

The control object is 1,476 KB, which is over twice as large as NRL2.DAT's 676 KB size. The test object for comparison is 216 KB, a deflation of ~1/3 the DAT size and ~1/7 the control size.

I tested the objects in-game using a debug build. When mousing over the control object, the game freezes for 1-3 seconds if the object is not added to the park or the object is not yet used in the park. When mousing over the test object, the game freezes for roughly 1/20 of a second.

I have not tested this, but I expect the 1-3 seconds of freezing that occur when mousing over the control object also occur when loading the game. Should many large objects (i.e. rides) be distributed in a single park, the 1-3 seconds can grow to a significant amount of time just to load objects. Beyond fixing the inconvenience of the game freezing when mousing over the object, the deflated file size is a boon for networking and future custom rides and the potentially-decreased loading time are major incentives to adding this functionality.

**Reply by spacek531** [reply]
@X123M3-256 if you have the chance, can you test the zip-compressed size and loading time of your raptor object using sprite atlas and using LGX?

**Comment by duncanspumpkin** [comment]
This is the opposite of what we want. Ideally rle should not be used anywhere it's just a slow way to handle transparency.

**Reply by spacek531** [reply]
I will have to verify this but I believe spritefile archives support images without rle

**Reply by IntelOrca** [reply]
@duncanspumpkin I tried converting the whole of g1.dat to raw format, and the performance was worse. I guess being able to skip large chunks of bytes is beneficial when it comes to bitbliting. Nevertheless, RLE is unnecessary for OpenGL.

The other way I wanted to investigate for speeding up load times, is to cache the data and image table when loading the object for the first time, so that the next time it loads, it will be quick. It would bypass even the time taken to deserialise the JSON.

**Comment by spacek531** [comment]
Spritefile archives do support both RLE and raw images. I've tested the branch with no modifications with @X123M3-256 's rmct1 object with atlas png, RLE spritefile, and raw spritefile. RLE and raw spritefiles saw similar performance - only one or two frames lost. Atlas png was a 1-2 second freeze. All tests were performed by moving the mouse over the object's entry in the object selection window.

Filesize of the atlas is much better than single images. The atlas parkobj provided by xtreme is 837 KB. The RLE parkobj is 915 KB. The raw parkobj is 726 KB. The compression algorithm works in mysterious ways, because the uncompressed RLE spritefile is larger than the uncompressed raw spritefile, with the json being the same size. Likewise, the uncompressed atlas images sum to less than 800 KB, so why the zip is larger, I have no idea.

The 3 files for testing:

[openrct2.ride.rmct1-lgx-raw.zip](https://github.com/OpenRCT2/OpenRCT2/files/8189422/openrct2.ride.rmct1-lgx-raw.zip)
[openrct2.ride.rmct1-lgx-rle.zip](https://github.com/OpenRCT2/OpenRCT2/files/8189423/openrct2.ride.rmct1-lgx-rle.zip)
[openrct2.ride.rmct1.zip](https://github.com/OpenRCT2/OpenRCT2/files/8189424/openrct2.ride.rmct1.zip)

**Reply by Gymnasiast** [reply]
ZIP works best for files that aren’t compressed already, so I’m not surprised that the non-RLE version compresses down to a smaller ZIP.
